keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,633 @@
+	// The grace period is the duration in seconds after the processes running in the pod are sent
+	// a termination signal and the time when the processes are forcibly halted with a kill signal.
+	// Set this value longer than the expected cleanup time for your process.
+	// If this value is nil, the pod's terminationGracePeriodSeconds will be used. Otherwise, this
+	// value overrides the value provided by the pod spec.
+	// Value must be non-negative integer. The value zero indicates stop immediately via
+	// the kill signal (no opportunity to shut down).
+	// This is a beta field and requires enabling ProbeTerminationGracePeriod feature gate.
+	// +optional
+	TerminationGracePeriodSeconds *int64
+}
+```
+
+### Existing Behavior
+
+Taking the Seconds fields, in order of the struct.
+InitialDelaySeconds
+TimeoutSeconds
+PeriodSeconds
+
+#### Defaulting logic
+ - InitialDelaySeconds, has no defaulting logic.
+   Therefore it would get the golang defaulting logic, and default to 0.
+ - TimeoutSeconds, defaults to 1 second if unset (which includes the explicitly set to 0 state). This is the last ending time before the timeout fails. Defaulting, thus, results in slow failures (when one considers a full second of time in cpu time).
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L224-L226
+
+```go
+ 	if obj.TimeoutSeconds == 0 {
+		obj.TimeoutSeconds = 1
+	}
+```
+
+ - Period Seconds is the biggest hurdle in charging and overall cost to the cluster, but also the most useful if set correctly.
+PeriodSeconds defaults to 10 seconds if unset (which includes the explicitly set to 0 state). Fast failure detections are thus blocked with existing defaults.
+https://github.com/kubernetes/kubernetes/blob/e19964183377d0ec2052d1f1fa930c4d7575bd50/pkg/apis/core/v1/defaults.go#L227-L229
+
+```go
+	if obj.PeriodSeconds == 0 {
+		obj.PeriodSeconds = 10
+	}
+```
+
+#### Validation of fields
+
+Must be non-negative, Zero or greater.
+```go
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.InitialDelaySeconds), fldPath.Child("initialDelaySeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.TimeoutSeconds), fldPath.Child("timeoutSeconds"))...)
+	allErrs = append(allErrs, ValidateNonnegativeField(int64(probe.PeriodSeconds), fldPath.Child("periodSeconds"))...)
+
+```
+
+#### Fields to Add
+What fields may be necessary to add?
+
+In pkg/apis/core/types.go
+
+```go
+
+type Probe struct {
+    ...
+    ...
+    // How often (in milliseconds) to offset PeriodSeconds when performing the probe.
+	// +optional
+	PeriodMilliseconds *int32
+	// Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated.
+	// +optional
+	InitialDelayMilliseconds *int32
+	// Length of time (in milliseconds) to offset TimeoutMilliseconds before health checking times out.
+	// +optional
+	TimeoutMilliseconds *int32
+}
+
+```
+#### Logic for Added Fields
```

Comment:
can this be restricted in validation to specific probes?

is it possible to only provide these higher granularities when a pod is not ready/live/startup, but once its in its happy state, we can backoff to a regular cadence?

Reply:
let's start alpha with:

Readiness and Startup supported for alpha, Liveness not supported in alpha 

type
====
exec(500ms minimum duration for one check drop back to the original second range/value after first check)
grpc(200ms minimum for one check drop back to the original second range)
http(200ms minimum for one check drop back to the original second range)

note for exec, grpc, http .. for startup and readiness will will support the millisecond range greater than zero.. in other words if they know 1.4 seconds is the right range and 1s just will not work they can ask for 1second plus 400mills vs having to go pick one of 1second where the first is known to fail or 2 where they know it's 600mills slower than it should be because 99% of runs will be less than 1.4s

Reply:
let's start alpha with:

Readiness and Startup supported for alpha, Liveness not supported in alpha 

type
====
exec(500ms minimum periodSeconds until success, then drop back to the original second range/value after first check)
grpc(200ms minimum until success, then drop back to the original second range)
http(200ms minimum until success, then drop back to the original second range)

note for exec, grpc, http .. for startup and readiness will will support the millisecond range greater than zero.. in other words if they know 1.4 seconds is the right range and 1s just will not work they can ask for 1second plus 400mills vs having to go pick one of 1second where the first is known to fail or 2 where they know it's 600mills slower than it should be because 99% of runs will be less than 1.4s   

keep initialDelayMilliseconds because zero is the current minimum..

the periodMilliseconds option for periodSeconds becomes a till success with appropriate minimums

will investigate timeoutMilliseconds for inclusion/exclusion based on validity of use cases (probably exclude)

Reply:
should be resolved now ... cheers

Reply:
These changes would be fine for our knative use case